### PR TITLE
dev-users-and-posts Pull Request

### PR DESCRIPTION
**Pull request by FSolM**
Third Milestone's Pull Request

### CODE REVIEW

File: app/models/post.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+class Post < ApplicationRecord
+  validates :author, presence: true
+  validates :content, presence: true
+  
+
+  # Relationship with User
+  belongs_to :author, foreign_key: :author_id, class_name: :User
```
Comment (mcrd25) [inline app/models/post.rb]: consider what `foreign_key: :author_id` implies

Comment (mcrd25) [reply app/models/post.rb]: 🤔

Comment (FSolM) [reply app/models/post.rb]: HAHAHAHA
Redundant, isn't it? 😅

File: app/controllers/posts_controller.rb [hunk 2]
```diff
@@ -0,0 +1,69 @@
+class PostsController < ApplicationController
+  def new
+    if signed_in?
+      @post = Post.new
```
Comment (mosaaleb) [inline app/controllers/posts_controller.rb]: I see that for every action we have conditions to check if the user is signed in or not. We can clean our posts controller by taking advantage of devise helper method [`authenticate_user!`](https://github.com/plataformatec/devise#controller-filters-and-helpers). That way we no longer have repeated code and have DRY'd our code.

File: app/models/post.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
+class Post < ApplicationRecord
+  validates :author, presence: true
+  validates :content, presence: true
+  
+
```
Comment (mosaaleb) [inline app/models/post.rb]: - **[OPTOINAL]** Unless you have a very strong reason to change the name of the model from `user` to `author`, keep it simple. `belongs_to :user` is much simpler and also is readable. Same for User model `has_many :posts, dependent: destroy`. You will have the time in the next milestone to play with `class_name` and `foreign_key`

File: app/models/user.rb [hunk 4]
```diff
@@ -2,5 +2,8 @@ class User < ApplicationRecord
   # Include default devise modules. Others available are:
   # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
   devise :database_authenticatable, :registerable,
-         :recoverable, :rememberable, :validatable
-end
+         :recoverable, :rememberable, :validatable, :omniauthable
+
```
Comment (mosaaleb) [inline app/models/user.rb]: We don't have any kind of validation in user model. Let's add presence validation for name and email.

File: config/routes.rb [hunk 5]
```diff
@@ -1,4 +1,24 @@
...
+                                    omniauth_callbacks: "user/omniauth_callbacks",
+                                  }
+
+  resources :users, only: [:show]
+  resources :posts, only: [:new, :create, :destroy, :edit, :update]
+
+  devise_scope :user do
+    authenticated :user do
+      root 'feed#index', as: :authenticated_root
+    end
```
Comment (mosaaleb) [inline config/routes.rb]: :clap:

File: app/controllers/posts_controller.rb [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+  end
+
+  private
+    def post_params
+      params.require(:post).permit(:content)
+    end
+
+    def patch_params
+      params.require(:patch).permit(:content)
+    end
```
[the remark this answers is not in the document]
Comment (mosaaleb) [reply app/controllers/posts_controller.rb]: `require` method takes the post model name which in our case `post`, it doesn't take the method as a parameter. In our case the params for update are the same params for create, so we can just use `post_params`. Please check [this](https://api.rubyonrails.org/classes/ActionController/Parameters.html) for more information.

**Comment by mosaaleb** [comment]
> I don't understand the second requirement, should we implement the test cases for commenting? I believe that belongs to the next milestone. Fixed everything else and submitted for another code review

Sorry for that I meant feature test for just posting.

File: app/models/post.rb (hunk 3 above)
Comment (mosaaleb) [reply app/models/post.rb]: Looks valid :+1:

File: app/models/user.rb [hunk 7]
```diff
@@ -2,5 +2,12 @@ class User < ApplicationRecord
...
+  validates :name, presence: true
+  validates :email, presence: true, format: { with: /\A([\w+\-].?)+@[a-z\d\-]+(\.[a-z]+)*\.[a-z]+\z/i,}
+  
+
```
Comment (mosaaleb) [inline app/models/user.rb]: If you had a look at `devise.rb` file you will find that devise already has a format for email and does email validations automatically, we don't have to make validations for email.

**Pull request merged**